### ISSUE TITLE
Falied reproducing MARTHE in CIFAR100...

### ISSUE THREAD

**Issue by YanwenZhu** [awslabs/adatune#3]
Hi, I was trying to reproduce your experiment in CIFAR100 but got into trouble, would you please give a little help?

1. When running the code like 
`python bin/rtho.py --network resnet --dataset cifar_100 --optimizer adam --lr 1e-5 --mu -0.1`
I cannot reproduce the test accuracy presented in your paper, which is over 75% after 200 epoch, and I got only 40%. The learning rate was descending from 1e-5 to around 1e-7 and some of the learning rates are less than 0. Setting `mu=0` couldn't help, and `mu=0.99` will result in the training loss being `nan`. How can I fix this?

2. Is computing hypergradient for learning rate supposed to be slow? I'm using a RTX 2080, and it takes more than 15 minutes to compute hypergradient( the `compute_hg` function) in one epoch. Is it because there are too many matrix multiplications or I didn't do it right?

I would really appreciate it if you can look into these problems!

**Comment by jmikko** [comment]
Hi @YanwenZhu 

Thank you for your interest in our project.
Let me try to help you.

1. With `mu=0` and `alpha=0` you can obtain the method from d Baydin et al. (2018) (_HD_ in our paper). You have to select the appropriate `beta` value. In this sense the performance should be similar to what reported both in our and their paper.
In the case of `mu=1` and `alpha=0` you can use the method from Franceschi et al. (2017) (_RTHO_ in our paper). This method is more sensitive to `beta`, and it requires a smaller value due to possible divergence (and consequently `nan` loss).
Finally, with `mu` between ]0,1[ and `alpha=0` you can obtain different interpolations between the two methods (and you would see that the optimal `mu` is usually not 0 nor 1, but something in the range ]0,1[). Also in this case you have to select the appropriate `beta`. In our experiments, `mu` close to 1 performs better than close to 0.
To conclude, if you use `mu` negative, you are asking to the method to try to find the best `mu` automatically, and with `alpha` positive (not zero), you are asking to the system to try to adapt the value of `beta` too. In this case you have to select `alpha` small enough to avoid divergence. A simple linear search works in this case.

2. I have not a RTX2080 to test with right now but I can say that computing the hypergradient is indeed the biggest computational burden of our method. Theoretically, our method should increase the computational complexity of 2x or 3x times the normal training. In practice, due to PyTorch bottlenecks, we noticed (on CIFAR100 with ResNet) an increase of 5x or 6x times the normal training.

I hope this helps. Feel free to reach me also at this email if you need: donini@amazon.com

Best,
Michele.

**Comment by YanwenZhu** [comment]
Thanks for your reply! I guess I didn't change  `alpha` and `beta` from their default values, in which case `alpha` is way too big for `beta`. I will re-run the experiment with more careful parameter selecting.

Thanks again for your time and patience!
